### PR TITLE
[IMP] payment_stripe: support Sofort

### PR DESCRIPTION
**Pull request by chevalierv**
Sofort enables customers in Austria, Belgium, Germany, Italy,
Netherlands and Spain to pay via direct online payouts from their bank
accounts. Customers log into Sofort's payment page to approve payments.

task-2577887

### CODE REVIEW

**Comment by danielosser** [comment]
Hi. 

Would it be possible to add Klarna for the Nordic region as well?

thanks in advance
Daniel 

> Sofort enables customers in Austria, Belgium, Germany, Italy, Netherlands and Spain to pay via direct online payouts from their bank accounts. Customers log into Sofort's payment page to approve payments.
> 
> task-2577887

**Comment by Feyensv** [comment]
> Hi.
> 
> Would it be possible to add Klarna for the Nordic region as well?

cc: @pierre-pde

File: addons/payment_stripe/controllers/main.py [hunk 1]
```diff
@@ -113,7 +117,9 @@ def stripe_webhook(self):
                         )
                         self._include_setup_intent_in_feedback_data(setup_intent, data)
                     # Handle the feedback data crafted with Stripe API objects as a regular feedback
-                    request.env['payment.transaction'].sudo()._handle_feedback_data('stripe', data)
+                    request.env['payment.transaction'].with_user(
+                        SUPERUSER_ID
+                    ).sudo()._handle_feedback_data('stripe', data)
```
Comment (Feyensv) [inline addons/payment_stripe/controllers/main.py:122]: why is this change necessary?
If it's for what I believe it is, shouldn't it be modified in the core payment code ? :thinking:

File: addons/sale/models/payment_transaction.py [hunk 2]
```diff
@@ -33,17 +33,22 @@ def _compute_sale_order_ids_nbr(self):
...
-
-            if record.acquirer_id.provider == 'transfer':
-                for so in record.sale_order_ids:
-                    so.reference = record._compute_sale_order_reference(so)
-            # send order confirmation mail
-            sales_orders._send_order_confirmation_mail()
+        # When using webhooks, this function can be called two times (when the user comes back and
+        # when the webhooks notify the db). To avoid duplicates, check if the transaction is not
+        # already pending
+        if premilinary_state != 'pending' and  self.state == 'pending':
```
Comment (Feyensv) [inline addons/sale/models/payment_transaction.py:42]: Seems like a fix to be done in stable, did you try reproducing it in stable version (let's say 15+) ?

**Comment by danielosser** [comment]
> > Hi.
> > Would it be possible to add Klarna for the Nordic region as well?
> 
> cc: @pierre-pde

I've made a pull request based on this PR with the additional Klarna Nordics additions. #81352 https://github.com/odoo/odoo/pull/81352 

//Daniel

**Review by Feyensv — COMMENTED** [review]
:eyes:

**Comment by danielosser** [comment]
> 

This code was originally done in to this branch: https://github.com/odoo-dev/odoo/tree/master-payment-stripe-support_sofort-vcr

I expanded the code to also handle Klarna.

I've tested this module with Klarna addition on 15 stable and it seem to work fine.

File: addons/payment_stripe/controllers/main.py (hunk 1 above)
Comment (chevalierv) [reply addons/payment_stripe/controllers/main.py:122]: Without the superuser_id, the log notes from the webhook are displayed as "Public User" instead of "OdooBot". I don't know if this should be modified in the core payment code but if you think it should, I'll do it.

File: addons/sale/models/payment_transaction.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (chevalierv) [reply addons/sale/models/payment_transaction.py:42]: probably not 🙄

Comment (chevalierv) [reply addons/sale/models/payment_transaction.py:42]: I think that this is specifically linked to refunds but I'll check it in stable version to be sure

File: addons/payment_stripe/controllers/main.py (hunk 1 above)
Comment (Feyensv) [reply addons/payment_stripe/controllers/main.py:122]: Might be a good separate commit to make sure message/log notes posted during transactions processing (from webhook or returns from redirects) are posted as 'Odoobot'.

Doesn't make sense for me to do that kind of change only in the custom stripe code.
Also, `with_user(superuser_id)` has multiple side effects, including being always sudoed, which might not be something we want to apply for the whole `_handle_feedback_data` call.  I really prefer if the user change is only applied in the targeted places where messages/log notes are posted.

File: addons/sale/models/payment_transaction.py (hunk 2 above)
Comment (chevalierv) [reply addons/sale/models/payment_transaction.py]: Currently, I don't see any cases in Stable where `_set_pending` could be triggered two times. But since we've added webhooks in a lot of other tasks and acquirers in master, I think that it may be a good idea to change that in a separate commit because it's not linked directly to Stripe

File: addons/sale/models/payment_transaction.py [hunk 3]
```diff
@@ -33,17 +33,24 @@ def _compute_sale_order_ids_nbr(self):
...
-            if record.acquirer_id.provider == 'transfer':
-                for so in record.sale_order_ids:
-                    so.reference = record._compute_sale_order_reference(so)
-            # send order confirmation mail
-            sales_orders._send_order_confirmation_mail()
+        # When using webhooks, this function can be called two times (when the user comes back and
+        # when the webhooks notify the db). To avoid duplicates, check if the transaction is not
+        # already pending
+        if previous_state != 'pending' and self.state == 'pending':
+            for tx in self:
```
Comment (Feyensv) [inline addons/sale/models/payment_transaction.py:43]: Either we support multiple tx's in `_set_pending` or not, but if you receive multiple txs, the line `previous_state = self.state` will crash :eyes:

File: addons/sale/models/payment_transaction.py [hunk 4]
```diff
@@ -36,17 +36,24 @@ def _compute_sale_order_ids_nbr(self):
...
-            if record.acquirer_id.provider == 'transfer':
-                for so in record.sale_order_ids:
-                    so.reference = record._compute_sale_order_reference(so)
-            # send order confirmation mail
-            sales_orders._send_order_confirmation_mail()
+        # When using webhooks, this function can be called two times (when the user comes back and
+        # when the webhooks notify the db). To avoid duplicates, check if the transaction is not
+        # already pending
+        for tx in self:
+            if previous_states[tx.id] != 'pending' and self.state == 'pending':
```
Comment (Feyensv) [inline addons/sale/models/payment_transaction.py:46]: for tx in self:
    self.state == 'pending'

:eyes:
(Not for this task, but this override of `_set_pending` seems quite crappy, probably a good place for further cleanups :D)
